src/map/script.c

```diff
@@ -15043,22 +15043,50 @@ BUILDIN(dispbottom)
  * All The Players Full Recovery
  * (HP/SP full restore and resurrect if need)
  *------------------------------------------*/
-BUILDIN(recovery)
-{
-	struct map_session_data *sd;
-	struct s_mapiterator* iter;
+int buildin_recovery_pc_sub(struct map_session_data *sd, va_list ap) {
+	nullpo_retr(0, sd);
 
-	iter = mapit_getallusers();
-	for (sd = BL_UCAST(BL_PC, mapit->first(iter)); mapit->exists(iter); sd = BL_UCAST(BL_PC, mapit->next(iter))) {
-		if(pc_isdead(sd))
-			status->revive(&sd->bl, 100, 100);
-		else
-			status_percent_heal(&sd->bl, 100, 100);
-		clif->message(sd->fd,msg_sd(sd,880)); // "You have been recovered!"
+	if (pc_isdead(sd)) {
+		status->revive(&sd->bl, 100, 100);
+	} else {
+		status_percent_heal(&sd->bl, 100, 100);
+	}
+
+	return 0;
+}
+
+int buildin_recovery_sub(struct block_list *bl, va_list ap) {
+	return script->buildin_recovery_pc_sub(BL_CAST(BL_PC, bl), ap);
+}
+
+BUILDIN(recovery) {
+	if (script_hasdata(st, 2)) {
+		if (script_isstringtype(st, 2)) {
+			int16 m = map->mapname2mapid(script_getstr(st, 2));
+
+			if (m == -1) {
+				ShowWarning("script:recovery: invalid map!\n");
+				return false;
+			}
+
+			if (script_hasdata(st, 6)) {
+				int16 x1 = script_getnum(st, 3);
+				int16 y1 = script_getnum(st, 4);
+				int16 x2 = script_getnum(st, 5);
+				int16 y2 = script_getnum(st, 6);
+				map->foreachinarea(script->buildin_recovery_sub, m, x1, y1, x2, y2, BL_PC);
+			} else {
+				map->foreachinmap(script->buildin_recovery_sub, m, BL_PC);
+			}
+		} else {
+			script->buildin_recovery_pc_sub(map->id2sd(script_getnum(st, 2)), NULL);
```

Comment:
It's better to split this into two parts: first retrieve the sd pointer (you can use `script->id2sd()`, which is a wrapper for `map->id2sd()` that already displays an error message on the console in case of failure), and then call the `script->buildin_recovery_pc_sub()` function only in case of success, in order to avoid the null pointer report in case of logged off character